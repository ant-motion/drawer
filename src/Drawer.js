import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import ContainerRender from 'rc-util/lib/ContainerRender';
import getScrollBarSize from 'rc-util/lib/getScrollBarSize';
import { dataToArray, transitionEnd, addEventListener, removeEventListener, transformArguments } from './utils';

const IS_REACT_16 = 'createPortal' in ReactDOM;

const currentDrawer = {};
let bodyAddTransitionEnd = false;
const windowIsUndefined = typeof window === 'undefined';
class Drawer extends React.PureComponent {
  static defaultProps = {
    prefixCls: 'drawer',
    placement: 'left',
    getContainer: 'body',
    level: 'all',
    duration: '.3s',
    ease: 'cubic-bezier(0.78, 0.14, 0.15, 0.86)',
    onChange: () => { },
    onMaskClick: () => { },
    onHandleClick: () => { },
    handler: (
      <div className="drawer-handle">
        <i className="drawer-handle-icon" />
      </div>
    ),
    showMask: true,
    maskStyle: {},
    wrapperClassName: '',
    className: '',
  };

  constructor(props) {
    super(props);
    this.levelDom = [];
    this.contextDom = null;
    this.maskDom = null;
    this.handlerdom = null;
    this.mousePos = null;
    this.firstEnter = false;// 记录首次进入.
    this.timeout = null;
    this.drawerId = Number((Date.now() + Math.random()).toString()
      .replace('.', Math.round(Math.random() * 9))).toString(16);

    if (props.onIconClick || props.parent || props.iconChild || props.width // eslint-disable-line react/prop-types
      || props.handleChild) {  // eslint-disable-line react/prop-types
      console.warn(// eslint-disable-line no-console
        'rc-drawer-menu API has been changed, please look at the releases, ' +
        'https://github.com/react-component/drawer-menu/releases'
      );
    }
    const open = props.open !== undefined ? props.open : !!props.defaultOpen;
    currentDrawer[this.drawerId] = open;
    this.state = {
      open,
    };
  }
  componentDidMount() {
    if (!windowIsUndefined) {
      if (!bodyAddTransitionEnd) {
        addEventListener(
          document.body,
          transitionEnd,
          this.onBodyTransitionEnd
        );
        bodyAddTransitionEnd = true;
      }
      let passiveSupported = false;
      window.addEventListener(
        'test',
        null,
        Object.defineProperty({}, 'passive', {
          get: () => {
            passiveSupported = true;
            return null;
          },
        })
      );
      this.passive = passiveSupported ? { passive: false } : false;
    }
    const open = this.getOpen();
    if (this.props.handler || open) {
      this.getDefault(this.props);
      this.forceUpdate();
    }
  }

  componentWillReceiveProps(nextProps) {
    const { open, placement } = nextProps;
    if (open !== undefined && open !== this.props.open) {
      this.isOpenChange = true;
      // 没渲染 dom 时，获取默认数据;
      if (!this.container) {
        this.getDefault(nextProps);
      }
      this.setState({
        open,
      });
    }
    if (placement !== this.props.placement) {
      // test 的 bug, 有动画过场，删除 dom
      this.contextDom = null;
    }
    if (this.props.level !== nextProps.level) {
      this.getParentAndLevelDom(nextProps);
    }
  }

  componentDidUpdate() {
    // dom 没渲染时，重走一遍。
    if (!this.firstEnter && this.container) {
      this.forceUpdate();
      this.firstEnter = true;
    }
  }

  componentWillUnmount() {
    if (this.container) {
      this.setLevelDomTransform(false, true);
      // 拦不住。。直接删除；
      if (this.props.getContainer) {
        this.container.parentNode.removeChild(this.container);
      }
    }
    if (!windowIsUndefined) {
      removeEventListener(
        document.body,
        transitionEnd,
        this.onBodyTransitionEnd
      );
    }
    clearTimeout(this.timeout);
    delete currentDrawer[this.drawerId];
    // suppport react15
    if (IS_REACT_16) {
      return;
    }
    this.renderComponent({
      afterClose: this.removeContainer,
      onClose() { },
      visible: false,
    });
  }

  onMaskTouchEnd = e => {
    this.props.onMaskClick(e);
    this.onTouchEnd(e, true);
  };

  onIconTouchEnd = e => {
    this.props.onHandleClick(e);
    this.onTouchEnd(e);
  };
  onTouchEnd = (e, close) => {
    if (this.props.open !== undefined) {
      return;
    }
    const open = close || this.state.open;
    this.isOpenChange = true;
    this.setState({
      open: !open,
    });
  };

  onBodyTransitionEnd = (e) => {
    if (e.target === e.currentTarget) {
      document.body.style.transform = '';
      document.body.style.transition = '';
      document.body.style.position = '';
      if (this.dom) {
        this.dom.style.transition = '';
      }
      if (!this.state.open) {
        document.body.style.overflow = '';
        if (this.props.placement === 'right' && this.maskDom) {
          this.maskDom.style.right = '';
          this.maskDom.style.width = '';
        }
      }
    }
  }

  getDefault = props => {
    this.getParentAndLevelDom(props);
    if (props.getContainer || props.parent) {
      this.container = this.defaultGetContainer();
    }
  };

  getContainer = () => {
    return this.container;
  };
  getParentAndLevelDom = props => {
    if (windowIsUndefined) {
      return;
    }
    const { level, getContainer } = props;
    this.levelDom = [];
    if (getContainer) {
      if (typeof getContainer === 'string') {
        const dom = document.querySelectorAll(getContainer)[0];
        this.parent = dom;
      }
      if (typeof getContainer === 'function') {
        this.parent = getContainer();
      }
      if (typeof getContainer === 'object' && getContainer instanceof window.HTMLElement) {
        this.parent = getContainer;
      }
    }
    if (!getContainer && this.container) {
      this.parent = this.container.parentNode;
    }
    if (level === 'all') {
      const children = Array.prototype.slice.call(this.parent.children);
      children.forEach(child => {
        if (child.nodeName !== 'SCRIPT' && child.nodeName !== 'STYLE' && child !== this.container) {
          this.levelDom.push(child);
        }
      });
    } else if (level) {
      dataToArray(level).forEach(key => {
        document.querySelectorAll(key).forEach(item => {
          this.levelDom.push(item);
        });
      });
    }
  };

  setLevelDomTransform = (open, openTransition, placementName, value) => {
    const { placement, levelMove, duration, ease, onChange } = this.props;
    if (!windowIsUndefined) {
      this.levelDom.forEach(dom => {
        if (this.isOpenChange || openTransition) {
          /* eslint no-param-reassign: "error" */
          dom.style.transition = `transform ${duration} ${ease}`;
          addEventListener(dom, transitionEnd, this.trnasitionEnd);
          let levelValue = open ? value : 0;
          if (levelMove) {
            const $levelMove = transformArguments(levelMove, { target: dom, open });
            levelValue = open ? $levelMove[0] : $levelMove[1] || 0;
          }
          const placementPos = placement === 'left' || placement === 'top' ? levelValue : -levelValue;
          dom.style.transform = placementPos ? `${placementName}(${placementPos}px)` : '';
        }
      });
      // 处理 body 滚动
      const eventArray = ['touchstart'];
      const domArray = [document.body, this.maskDom, this.handlerdom, this.contextDom];
      const right = getScrollBarSize();
      const widthTransition = `width ${duration} ${ease}`;
      const trannsformTransition = `transform ${duration} ${ease}`;
      if (open && document.body.style.overflow !== 'hidden') {
        document.body.style.overflow = 'hidden';
        if (right) {
          document.body.style.position = 'relative';
          document.body.style.width = `calc(100% - ${right}px)`;
          document.body.style.transition = 'none';
          this.dom.style.transition = 'none';
          switch (placement) {
            case 'right':
              this.dom.style.transform = `translateX(-${right}px)`;
              break;
            case 'top':
            case 'bottom':
              this.dom.style.width = `calc(100% - ${right}px)`;
              break;
            default:
              break;
          }
          clearTimeout(this.timeout);
          this.timeout = setTimeout(() => {
            document.body.style.transition = widthTransition;
            this.dom.style.transition = `${trannsformTransition},${widthTransition}`;
            document.body.style.width = '';
            this.dom.style.width = '';
            this.dom.style.transform = '';
          });
        }
        // 手机禁滚
        if (document.body.addEventListener) {
          domArray.forEach((item, i) => {
            if (!item) {
              return;
            }
            item.addEventListener(
              eventArray[i] || 'touchmove',
              i ? this.removeMoveHandler : this.removeStartHandler,
              this.passive
            );
          });
        }
      } else if (!Object.keys(currentDrawer).some(key => currentDrawer[key])) {
        document.body.style.overflow = '';
        if (this.isOpenChange && right) {
          document.body.style.position = 'relative';
          document.body.style.overflowX = 'hidden'
          document.body.style.transition = 'none';
          document.body.style.width = `calc(100% + ${right}px)`;
          this.dom.style.transition = 'none';
          switch (placement) {
            case 'right': {
              this.dom.style.transform = `translateX(${right}px)`;
              this.maskDom.style.right = `${right}px`;
              this.maskDom.style.width = `calc(100% + ${right}px)`;
              break;
            }
            case 'top':
            case 'bottom': {
              this.dom.style.width = `calc(100% + ${right}px)`;
              break;
            }
            default:
              break;
          }
          clearTimeout(this.timeout);
          this.timeout = setTimeout(() => {
            document.body.style.transition = widthTransition;
            this.dom.style.transition = `${trannsformTransition},${widthTransition}`;
            document.body.style.width = '';
            this.dom.style.transform = '';
            this.dom.style.width = '';
          });
        }
        if (document.body.removeEventListener) {
          domArray.forEach((item, i) => {
            if (!item) {
              return;
            }
            item.removeEventListener(
              eventArray[i] || 'touchmove',
              i ? this.removeMoveHandler : this.removeStartHandler,
              this.passive
            );
          });
        }
      }
    }

    if (onChange && this.isOpenChange && this.firstEnter) {
      onChange(open);
      this.isOpenChange = false;
    }
  };

  getChildToRender = open => {
    const {
      className,
      prefixCls,
      style,
      placement,
      children,
      handler,
      showMask,
      maskStyle,
    } = this.props;
    const wrapperClassname = classnames(prefixCls, {
      [`${prefixCls}-${placement}`]: true,
      [`${prefixCls}-open`]: open,
      [className]: !!className,
    });
    const value = this.contextDom
      ? this.contextDom.getBoundingClientRect()[
        placement === 'left' || placement === 'right' ? 'width' : 'height'
      ] : 0;
    const placementName = `translate${placement === 'left' || placement === 'right' ? 'X' : 'Y'}`;
    // 百分比与像素动画不同步，第一次打用后全用像素动画。
    const defaultValue = !this.contextDom ? '100%' : `${value}px`;
    const placementPos =
      placement === 'left' || placement === 'top' ? `-${defaultValue}` : defaultValue;
    const transform = open ? '' : `${placementName}(${placementPos})`;
    if (this.isOpenChange === undefined || this.isOpenChange) {
      this.setLevelDomTransform(open, false, placementName, value);
    }
    const handlerCildren = handler && React.cloneElement(handler, {
      onClick: (e) => {
        if (handler.props.onClick) {
          handler.props.onClick();
        }
        this.onIconTouchEnd(e);
      },
      ref: (c) => {
        this.handlerdom = c;
      }
    });
    return (
      <div
        className={wrapperClassname}
        style={style}
        ref={c => { this.dom = c; }}
      >
        {showMask && (
          <div
            className={`${prefixCls}-mask`}
            onClick={this.onMaskTouchEnd}
            style={maskStyle}
            ref={c => {
              this.maskDom = c;
            }}
          />
        )}
        <div
          className={`${prefixCls}-content-wrapper`}
          style={{ transform }}
        >
          <div
            className={`${prefixCls}-content`}
            ref={c => {
              this.contextDom = c;
            }}
            onTouchStart={open ? this.removeStartHandler : null} // 跑用例用
            onTouchMove={open ? this.removeMoveHandler : null} // 跑用例用
          >
            {children}
          </div>
          {handlerCildren}
        </div>
      </div>
    );
  };

  getOpen = () => (
    this.props.open !== undefined ? this.props.open : this.state.open
  )

  removeStartHandler = e => {
    if (e.touches.length > 1) {
      return;
    }
    this.startPos = {
      x: e.touches[0].clientX,
      y: e.touches[0].clientY,
    };
  };

  removeMoveHandler = e => {
    if (e.changedTouches.length > 1) {
      return;
    }
    const currentTarget = e.currentTarget;
    const differX = e.changedTouches[0].clientX - this.startPos.x;
    const differY = e.changedTouches[0].clientY - this.startPos.y;
    if (
      currentTarget === this.maskDom ||
      currentTarget === this.handlerdom ||
      (currentTarget === this.contextDom &&
        ((((currentTarget.scrollTop + currentTarget.offsetHeight >= currentTarget.scrollHeight &&
          differY < 0) ||
          (currentTarget.scrollTop <= 0 && differY > 0)) &&
          Math.max(Math.abs(differX), Math.abs(differY)) === differY) ||
          (((currentTarget.scrollLeft + currentTarget.offsetWidth >= currentTarget.scrollWidth &&
            differX < 0) ||
            (currentTarget.scrollLeft <= 0 && differX > 0)) &&
            Math.max(Math.abs(differX), Math.abs(differY)) === differX)))
    ) {
      e.preventDefault();
    }
  };

  trnasitionEnd = e => {
    removeEventListener(e.target, transitionEnd, this.trnasitionEnd);
    e.target.style.transition = '';
  };

  defaultGetContainer = () => {
    if (windowIsUndefined) {
      return null;
    }
    const container = document.createElement('div');
    this.parent.appendChild(container);
    if (this.props.wrapperClassName) {
      container.className = this.props.wrapperClassName;
    }
    return container;
  };

  render() {
    const { getContainer, wrapperClassName } = this.props;
    const open = this.getOpen();
    currentDrawer[this.drawerId] = open ? this.container : open;
    const children = this.getChildToRender(this.firstEnter ? open : false);
    if (!getContainer) {
      return (
        <div
          className={wrapperClassName}
          ref={c => {
            this.container = c;
          }}
        >
          {children}
        </div>
      );
    }
    if (!this.container || !open && !this.firstEnter) {
      return null;
    }
    // suppport react15
    if (!IS_REACT_16) {
      return (
        <ContainerRender
          parent={this}
          visible
          autoMount
          autoDestroy={false}
          getComponent={this.getChildToRender}
          getContainer={this.getContainer}
        >
          {({ renderComponent, removeContainer }) => {
            this.renderComponent = renderComponent;
            this.removeContainer = removeContainer;
            return null;
          }}
        </ContainerRender>
      );
    }
    return ReactDOM.createPortal(children, this.container);
  }
}

Drawer.propTypes = {
  wrapperClassName: PropTypes.string,
  className: PropTypes.string,
  children: PropTypes.node,
  style: PropTypes.object,
  defaultOpen: PropTypes.bool,
  open: PropTypes.bool,
  prefixCls: PropTypes.string,
  placement: PropTypes.string,
  level: PropTypes.oneOfType([PropTypes.string, PropTypes.array]),
  levelMove: PropTypes.oneOfType([PropTypes.number, PropTypes.func, PropTypes.array]),
  ease: PropTypes.string,
  duration: PropTypes.string,
  getContainer: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object, PropTypes.bool]),
  handler: PropTypes.any,
  onChange: PropTypes.func,
  onMaskClick: PropTypes.func,
  onHandleClick: PropTypes.func,
  showMask: PropTypes.bool,
  maskStyle: PropTypes.object,
};

export default Drawer;
