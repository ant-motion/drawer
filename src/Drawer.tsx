import * as React from 'react';
import Portal from '@rc-component/portal';
import type { PortalProps } from '@rc-component/portal';
import useLayoutEffect from 'rc-util/lib/hooks/useLayoutEffect';
import DrawerPopup from './DrawerPopup';
import type { DrawerPopupProps } from './DrawerPopup';
import { warnCheck } from './util';

export type Placement = 'left' | 'top' | 'right' | 'bottom';

export interface DrawerProps
  extends Omit<DrawerPopupProps, 'prefixCls' | 'inline' | 'scrollLocker'> {
  prefixCls?: string;
  open?: boolean;
  onClose?: (e: React.MouseEvent | React.KeyboardEvent) => void;
  destroyOnClose?: boolean;
  getContainer?: PortalProps['getContainer'];
}

const Drawer: React.FC<DrawerProps> = props => {
  const {
    open = false,
    prefixCls = 'rc-drawer',
    placement = 'right' as Placement,
    autoFocus = true,
    keyboard = true,
    width = 378,
    mask = true,
    maskClosable = true,
    getContainer,
    forceRender,
    afterOpenChange,
    destroyOnClose,
  } = props;

  const [animatedVisible, setAnimatedVisible] = React.useState(false);

  // ============================= Warn =============================
  if (process.env.NODE_ENV !== 'production') {
    warnCheck(props);
  }

  // ============================= Open =============================
  const [mounted, setMounted] = React.useState(false);

  useLayoutEffect(() => {
    setMounted(true);
  }, []);

  const mergedOpen = mounted ? open : false;

  // ============================ Focus =============================
  const panelRef = React.useRef<HTMLDivElement>();

  const lastActiveRef = React.useRef<HTMLElement>();
  useLayoutEffect(() => {
    if (mergedOpen) {
      lastActiveRef.current = document.activeElement as HTMLElement;
    }
  }, [mergedOpen]);

  // ============================= Open =============================
  const internalAfterOpenChange: DrawerProps['afterOpenChange'] =
    nextVisible => {
      setAnimatedVisible(nextVisible);
      afterOpenChange?.(nextVisible);

      if (
        !nextVisible &&
        lastActiveRef.current &&
        !panelRef.current?.contains(lastActiveRef.current)
      ) {
        lastActiveRef.current?.focus();
      }
    };

  // ============================ Render ============================
  if (!forceRender && !animatedVisible && !mergedOpen && destroyOnClose) {
    return null;
  }

  const drawerPopupProps = {
    ...props,
    open: mergedOpen,
    prefixCls,
    placement,
    autoFocus,
    keyboard,
    width,
    mask,
    maskClosable,
    inline: getContainer === false,
    afterOpenChange: internalAfterOpenChange,
    ref: panelRef,
  };

  return (
    <Portal
      open={mergedOpen || forceRender || animatedVisible}
      autoDestroy={false}
      getContainer={getContainer}
      autoLock={mask && (mergedOpen || animatedVisible)}
    >
      <DrawerPopup {...drawerPopupProps} />
    </Portal>
  );
};

if (process.env.NODE_ENV !== 'production') {
  Drawer.displayName = 'Drawer';
}

export default Drawer;
