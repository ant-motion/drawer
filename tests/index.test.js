/* eslint-disable */
import React from 'react';
import PropTypes from 'prop-types';
import { mount } from 'enzyme';
import Drawer from '../src';

function Div(props) {
  return (
    <div className="div-wrapper">
      {props.show && <Drawer wrapperClassName="drawer-wrapper" defaultOpen />}
    </div>
  );
}
Div.propTypes = {
  show: PropTypes.bool,
};

function DrawerComp(props) {
  return (
    <div className="react-wrapper">
      <div id="a" style={{ position: 'absolute', top: 0, left: 0 }}>test1</div>
      <Drawer getContainer={null} open={props.open} wrapperClassName="drawer-wrapper" />
    </div>
  );
}

function createClientXY(x, y) {
  return { clientX: x, clientY: y };
}

function createStartTouchEventObject({ x = 0, y = 0 }) {
  return { touches: [createClientXY(x, y)] };
}

function createMoveTouchEventObject({ x = 0, y = 0 }) {
  return { touches: [createClientXY(x, y)], changedTouches: [createClientXY(x, y)] };
}

describe('rc-drawer-menu', () => {
  let instance;
  it('single drawer', () => {
    instance = mount(<Drawer onIconClick={() => { }} />);
    const drawer = instance.find('.drawer');
    const drawerContent = instance.find('.drawer-content-wrapper');
    expect(!!drawer).toBe(true);
    expect(drawer.instance().parentNode.parentNode.tagName).toBe('BODY');
    console.log('clientWidth:', drawerContent.instance().clientWidth);
    expect(drawerContent.instance().style.transform).toBe('translateX(-100%)');
  });

  it('icon child is element', () => {
    instance = mount(<Drawer
      handler={<div className="a"><i >a</i></div>}
      level={null}
    />);
    const icon = instance.find('.a');
    console.log('icon.length:', icon.length);
    expect(icon.length).toBe(1);
  });

  it('default open drawer', () => {
    instance = mount(<Drawer
      handler={<i className="a">a</i>}
      defaultOpen
      level={[]}
    />);
    const drawer = instance.find('.drawer-content-wrapper').instance();
    const content = instance.childAt(0).childAt(1).childAt(0);
    content.simulate('touchStart',
      createStartTouchEventObject({ x: 100, y: 0 }));
    content.simulate('touchMove',
      createMoveTouchEventObject({ x: 150, y: 10 }));
    content.simulate('touchEnd',
      createMoveTouchEventObject({ x: 200, y: 0 }));
    content.simulate('touchStart',
      createStartTouchEventObject({ x: 0, y: 0 }));
    content.simulate('touchMove',
      createMoveTouchEventObject({ x: 0, y: 10 }));
    content.simulate('touchEnd',
      createMoveTouchEventObject({ x: 0, y: 10 }));
    console.log('transform is empty:', drawer.style.transform);
    expect(drawer.style.transform).toEqual('');
  });

  it('handler is null，open=true', () => {
    const instance = mount(<Drawer
      handler={false}
      open
      level={null}
    />);
    expect(instance.render()).toMatchSnapshot();
  });
  it('handler is null，open=false', () => {
    const instance = mount(<Drawer
      handler={false}
      open={false}
      level={null}
    />);
    expect(instance.render()).toMatchSnapshot();
  });
  it('switch open drawer', () => {
    instance = mount(<DrawerComp />);
    jest.useFakeTimers();
    const drawer = instance.find('.drawer-content-wrapper').instance();
    console.log('第一次：', drawer.style.transform);
    expect(drawer.style.transform).toBe('translateX(-0px)');
    instance.setProps({
      open: true,
    });
    jest.runAllTimers();
    console.log(drawer.style.transform);
    expect(drawer.style.transform).toEqual('');
    instance.setProps({
      open: false,
    });
    console.log('第二次：', drawer.style.transform);
    jest.runAllTimers();
    expect(drawer.style.transform).toBe('translateX(-0px)');
    jest.useRealTimers();
  });
  it('getContainer is null', () => {
    instance = mount(<div className="react-wrapper">
      <div id="a" className="a" style={{ position: 'absolute', top: 0, left: 0 }}>test1</div>
      <Drawer
        getContainer={null}
        defaultOpen
        level="#a"
        wrapperClassName="drawer-wrapper"
      />
    </div>);
    const drawer = instance.find('.drawer').instance();
    const a = instance.find('#a').instance();
    console.log('a transform:', a.style.transform);
    expect(a.style.transform).toBe('');
    console.log(drawer.parentNode.parentNode.className);
    expect(drawer.parentNode.className).toBe('drawer-wrapper');
    expect(drawer.parentNode.parentNode.className).toBe('react-wrapper');
  });
  it('click open close', () => {
    instance = mount(<Drawer level="b" levelMove={200} />);
    const content = instance.find('.drawer-content-wrapper').instance();
    console.log(content.style.transform);
    expect(content.style.transform).toBe('translateX(-100%)');
    const handle = instance.find('.drawer-handle');
    handle.simulate('click');
    console.log(content.style.transform);
    expect(content.style.transform).toBe('');
    const mask = instance.find('.drawer-mask');
    mask.simulate('click');
    console.log(content.style.transform);
    expect(content.style.transform).toBe('translateX(-0px)');
  });
  it('will unmount', () => {
    instance = mount(<Div show />);
    const divWrapper = instance.find('.div-wrapper').instance();
    const content = instance.find('.drawer-content-wrapper').instance();
    console.log(content.style.transform);
    expect(content.style.transform).toBe('');
    instance.setProps({
      show: false,
    });
    expect(divWrapper.children.length).toBe(0);
  });
  it('placement change', () => {
    instance = mount(<Drawer level={null} />);
    const content = instance.find('.drawer-content-wrapper').instance();
    console.log(content.style.transform);
    expect(content.style.transform).toBe('translateX(-100%)');
    instance.setProps({
      placement: 'top',
      level: ['a', 'b'],
    });
    console.log(content.style.transform);
    expect(content.style.transform).toBe('translateY(-100%)');
  });
  it('levelMove is Array', () => {
    instance = mount(<Drawer handler={false} levelMove={[200, 0]} />);
    expect(instance.children().length).toBe(0);
    instance.setProps({
      open: true,
    });
    instance.setProps({
      open: false,
      levelMove: () => (200),
    });
    instance.setProps({
      open: true,
      levelMove: [200],
    });
  });
  it('handler is null, render is null', () => {
    instance = mount(<Drawer handler={false} levelMove={200} />);
    console.log(instance.children())
    expect(instance.children().length).toBe(0);
    instance.setProps({
      open: true,
    });
    console.log(instance.children())
    expect(instance.children().length).toBe(1);
  });
});
