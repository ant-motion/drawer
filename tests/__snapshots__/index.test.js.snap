// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rc-drawer-menu handler is null，open=false 1`] = `null`;

exports[`rc-drawer-menu handler is null，open=true 1`] = `
<div
  class="drawer drawer-left drawer-open"
>
  <div
    class="drawer-mask"
  />
  <div
    class="drawer-content-wrapper"
    style=""
  >
    <div
      class="drawer-content"
    />
  </div>
</div>
`;
